OpenSim/Common/Storage.cpp

```diff
@@ -123,44 +163,74 @@ Storage::Storage(int aCapacity,const string &aName) :
+Storage::Storage(const string &fileName, bool readHeadersOnly) :
+    StorageInterface(fileName),
     _storage(StateVector())
 {
     // SET NULL STATES
     setNull();
 
     // OPEN FILE
-    std::unique_ptr<ifstream> fp{IO::OpenInputFile(aFileName)};
+    std::unique_ptr<ifstream> fp{IO::OpenInputFile(fileName)};
     OPENSIM_THROW_IF(fp == nullptr, Exception,
-            "Storage: Failed to open file " + aFileName);
-
-    int nr=0,nc=0;
-    if (!parseHeaders(*fp, nr, nc) && _fileVersion <= 1) {
-        // Version 2 Storage files do not need nColumns and nRows in the
-        // header.
-        OPENSIM_THROW(Exception, "Storage: Failed to parse headers of file "
-                + aFileName);
-    }
-    if(_fileVersion > 1) {
-        OPENSIM_THROW_IF(readHeadersOnly, Exception, 
-                "Cannot read headers only if STO version is greater than 1.");
-        TimeSeriesTable table = STOFileAdapter::read(aFileName);
-        convertTableToStorage(table, *this);
-        return;
+            "Storage: Failed to open file '" + fileName + 
+            "'. Verify that the file exists at the specified location." );
+
+    bool isMotFile = SimTK::String::toLower(fileName).rfind(".mot") != string::npos;
+    bool isStoFile = SimTK::String::toLower(fileName).rfind(".sto") != string::npos;
+    bool useFileAdpater = true;
+
+    int nr = 0, nc = 0;
+
+    if (isMotFile || isStoFile) {
+        parseHeaders(*fp, nr, nc);
+        if (_fileVersion <= 1) { // If an old .sto or .mot format
+            // Must have valid number of rows and columns
+            OPENSIM_THROW_IF(nr < 1 && nc < 1, Exception,
+                "Storage: Failed to parse headers of file "
+                + fileName);
+            // Checks out as a valid old format, so use legacy code to process
+            useFileAdpater = false;
+        } 
+    }
+
+    if (useFileAdpater) { // For new .sto files and others that are not .mot
+        try {
+            // Try using FileAdpater to read all file types
+            OPENSIM_THROW_IF(readHeadersOnly, Exception,
+                "Cannot read headers only if not a STO file or its "
+                "version is greater than 1.");
+            FileAdapter::OutputTables tables = FileAdapter::readFile(fileName);
+            if (tables.size() > 1) {
+                cout << "Storage: cannot read data files with multiple tables. "
+                    << "Only the first table '" << tables.begin()->first << "' will "
+                    << "loaded as Storage." << endl;
+            }
+            convertTableToStorage(tables.begin()->second.get(), *this);
+            return;
+        }
+        catch (const std::exception& x) {
+            cout << "Storage: FileAdpater failed to read data file.\n"
+                << x.what() << endl;
+            if (isStoFile) 
+                cout << "Reverting to use conventional Storage reader." << endl;
+            else 
+                throw x;
```

Comment:
Can you add comments here about the specific scenarios you are expecting? For example, why what type of storage file can be read using the conventional reader that can't be read with the adapters?

Reply:
I do not anticipate the `FileAdapter` to fail on new files nor do I have any specific expectations that using the conventional reader will help. I only figure if it's a `.sto` file it might be worth trying anyways, in the unexpected case the `FileAdapter` fails.